src/lib/backend_fapi.c

```diff
@@ -727,3 +639,225 @@ CK_RV backend_fapi_rm_tobject(token *t, tobject *tobj) {
+    }
+
+    char label[sizeof(tok->label) + 1]; /* token-label length plus \0, cannot overflow */
+    label[sizeof(tok->label)] = '\0';
+    memcpy(&label[0], &tok->label[0], sizeof(tok->label));
+
+    /* FAPI may return the description (which is the label) or the path.
+       Thus we register our auth value for either. */
+    struct authtable authtable[] = {
+        { path, (char *)sealobjauth },
+        { &label[0], (char *)sealobjauth },
+        { NULL, NULL } };
+
+    rc = Fapi_SetAuthCB(tok->fapi.ctx, auth_cb, &authtable[0]);
+    if (rc) {
+        twist_free(sealobjauth);
+        LOGE("Fapi_SetAuthCB failed.");
+        goto error;
+    }
+
+    uint8_t *data;
+    size_t size;
+    rc = Fapi_Unseal(tok->fapi.ctx, path, &data, &size);
+    //TODO: Reenable once we switch to FAPI 3.0
+    //Fapi_SetAuthCB(tok->fapi.ctx, NULL, NULL);
+    twist_free(sealobjauth);
+    if (user && rc == TSS2_FAPI_RC_PATH_NOT_FOUND) {
+        rv = CKR_USER_PIN_NOT_INITIALIZED;
+        goto error;
+    }
+    if (rc) {
+        LOGE("Fapi_Unseal failed.");
+        goto error;
+    }
+
+    twist wrappingkeyhex = twistbin_new(data, size);
+    Fapi_Free(data);
+    if (!wrappingkeyhex) {
+        rv = CKR_HOST_MEMORY;
+        goto error;
+    }
+
+    if (tok->wrappingkey) {
+        twist_free(wrappingkeyhex);
+    } else {
+        tok->wrappingkey = twistbin_unhexlify(wrappingkeyhex);
+        twist_free(wrappingkeyhex);
+        if (!tok->wrappingkey) {
+            LOGE("Expected internal wrapping key in base 16 format");
+            goto error;
+        }
+    }
+
+    free(path);
+
+    /* Since tobject are esysdb backed, they need an active session. */
+    if (!tpm_session_active(tok->tctx)) {
+        LOGV("token parent object handle is 0x%08x", tok->pobject.handle);
+        CK_RV tmp = tpm_session_start(tok->tctx, tok->pobject.objauth, tok->pobject.handle);
+        if (tmp != CKR_OK) {
+            LOGE("Could not start Auth Session with the TPM.");
+            return tmp;
+        }
+    }
+
+    return CKR_OK;
+
+error:
+    free(path);
+    return rv;
+}
+
+CK_RV backend_fapi_token_changeauth(token *tok, bool user, twist toldpin, twist tnewpin) {
+
+    CK_RV rv = CKR_GENERAL_ERROR;
+    TSS2_RC rc;
+
+    char *path = tss_path_from_id(tok->id, user ? "usr":"so");
+    if (!path) {
+        LOGE("No path constructed.");
+        return CKR_HOST_MEMORY;
+    }
+
+    twist newsalthex = NULL;
+    twist newauthhex = NULL;
+    twist oldauth = NULL;
+
+    rv = utils_setup_new_object_auth(tnewpin, &newauthhex, &newsalthex);
+    if (rv != CKR_OK) {
+        goto out;
+    }
+    rv = CKR_GENERAL_ERROR;
+
+    oldauth = utils_hash_pass(toldpin, user ? tok->fapi.userauthsalt : tok->fapi.soauthsalt);
+    if (!oldauth) {
+        goto out;
+    }
+
+    char label[sizeof(tok->label) + 1]; /* token-label length plus \0, cannot overflow */
+    label[sizeof(tok->label)] = '\0';
+    memcpy(&label[0], &tok->label[0], sizeof(tok->label));
```

Comment:
labels need to be binary safe so if your banking on a null \0, you might not get what you want if one is embedded before that.

Reply:
Are you certain ? IMHO label might not be \0 terminated, but if there is a \0 then that is the terminator.
That's why I get the label and append a \0 to guarantee being a string...

Reply:
> Are you certain ? IMHO label might not be \0 terminated, but if there is a \0 then that is the terminator.
> That's why I get the label and append a \0 to guarantee being a string...
You guarantee that it's null terminated, nothing more. However, the data passed is supposed to an rfc2279 string thus UTF8 compatible, but we never check. We might want to add a check in init token, that it has no embedded NULL's then we're good.